osdk/tests/examples_in_book/test_and_run_projects.rs

```diff
@@ -48,16 +61,28 @@ fn create_and_test_library() {
         fs::remove_dir_all(&module_dir).unwrap();
     }
 
-    let mut command = cargo_osdk(&["new", module_name]);
-    command.current_dir(work_dir);
-    command.ok().unwrap();
+    let mut new_command = cargo_osdk(&["new", module_name]);
+    new_command.current_dir(work_dir);
+    new_command.ok().unwrap();
 
     let manifest_path = module_dir.join("Cargo.toml");
-    depends_on_local_ostd(manifest_path);
-
-    let mut command = cargo_osdk(&["test"]);
-    command.current_dir(&module_dir);
-    command.ok().unwrap();
+    depends_on_local_ostd(manifest_path.clone());
+
+    let mut opt_args = Vec::new();
+    let tdx_enabled = is_tdx_enabled();
+    if tdx_enabled {
+        let osdk_path = module_dir.join("OSDK.toml");
+        add_tdx_scheme_to_manifest(&osdk_path).unwrap();
+        opt_args.push("--scheme");
+        opt_args.push("tdx");
```

Comment:
These block of code is repeated four times across the test files. So is the `std::iter::once("command name").chain(...)` pattern. Such duplicated code is highly undesirable. And it exposes low-level details that should not appear in the test case itself, whose logic should be focused on the workflow, not how the command is formed.

The original version of `create_and_run_kernel` is short and sweet, around 30 lines of code. After being extended with TDX support, the function is bloated to 40 lines. But in fact, I think it is possible to almost leave the test case intact by modify the `util.rs`.

```rust
pub fn cargo_osdk<T: AsRef<OsStr>, I: IntoIterator<Item = T>>(args: I) -> Command {
    let args = if is_tdx_enabled() {
        extend_osdk_command_for_tdx(args)
    };

    let mut command = Command::cargo_bin("cargo-osdk").unwrap();
    command.arg("osdk");
    command.args(args);
    command
}

pub fn edit_manifest(manifest_path: impl AsRef<Path>) {
    if is_tdx_enabled() {
        add_tdx_scheme(manifest_path);
    };
    depends_on_local_ostd(manifest_path);
}

// This function is now private
fn depends_on_local_ostd(manifest_path: impl AsRef<Path>) {
    // Content not changed
}
```

With this new utility module, the original `create_and_run_kernel` function only needs one line changed, without adding more lines.

```rust
#[test]
fn create_and_run_kernel() {
    let work_dir = "/tmp";
    let os_name = "myos";

    let os_dir = PathBuf::from(work_dir).join(os_name);

    if os_dir.exists() {
        fs::remove_dir_all(&os_dir).unwrap();
    }

    let mut command = cargo_osdk(&["new", "--kernel", os_name]);
    command.current_dir(work_dir);
    command.ok().unwrap();

    // Makes the kernel depend on local OSTD
    let manifest_path = os_dir.join("Cargo.toml");
    edit_manifest(&manifest_path); // ONLY THIS LINE IS CHANGED!

    let mut command = cargo_osdk(&["build"]);
    command.current_dir(&os_dir);
    command.ok().unwrap();

    let mut command = cargo_osdk(&["run"]);
    command.current_dir(&os_dir);
    let output = command.output().unwrap();

    let stdout = String::from_utf8_lossy(&output.stdout).to_string();
    assert!(stdout.contains("Hello world from guest kernel!"));

    fs::remove_dir_all(&os_dir).unwrap();
}
```

Reply:
Learned a lot. Thanks! The slight differences are:
1. The `edit_manifest` function needs to modify the `Cargo.toml` and the `OSDK.toml`, so the function interface is modified.
2. When building the workspace example, the situation is a little complicated. We need to add the tdx scheme to the `OSDK.toml` file in `/tmp/myworkspace/`, and also need to modify the osdk dependencies in `/tmp/myworkspace/myos/Cargo.toml` and `/tmp/myworkspace/mylib/Cargo.toml`, so I handle them separately.